astropy/io/fits/tests/test_table.py

```diff
@@ -3270,6 +3270,72 @@ def test_empty_vla_raw_data(self):
                 [np.array([], dtype=np.int32), np.array([], dtype=np.int32)],
             )
 
+    def test_multidim_VLA_tables(self):
+        """
+        Check if multidimensional VLF are correctly write and read.
+        See https://github.com/astropy/astropy/issues/12860
+        and https://github.com/astropy/astropy/issues/7810
+        """
+        a = np.arange(5).reshape((5, 1))
+        b = np.arange(7).reshape((7, 1))
+        array = np.array([a, b], dtype=object)
+        col = fits.Column(name="test", format="PD(7)", dim="(1,7)", array=array)
+        fits.BinTableHDU.from_columns([col]).writeto(self.temp("test.fits"))
+
+        with fits.open(self.temp("test.fits")) as hdus:
+            assert hdus[1].columns.formats == ["PD(7)"]
+            np.array_equal(
+                hdus[1].data["test"],
+                [
+                    np.array([[0.0, 1.0, 2.0, 3.0, 4.0]]),
+                    np.array([[0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0]]),
+                ],
+            )
+
+        a = np.arange(10).reshape((5, 2))
+        b = np.arange(14).reshape((7, 2))
+        array = np.array([a, b], dtype=object)
+        col = fits.Column(name="test", format="PD(14)", dim="(2,7)", array=array)
+        fits.BinTableHDU.from_columns([col]).writeto(self.temp("test2.fits"))
+
+        with fits.open(self.temp("test2.fits")) as hdus:
+            assert hdus[1].columns.formats == ["PD(14)"]
+            np.array_equal(
+                hdus[1].data["test"],
+                [
+                    np.array(
+                        [[0.0, 1.0], [2.0, 3.0], [4.0, 5.0], [6.0, 7.0], [8.0, 9.0]]
+                    ),
+                    np.array(
+                        [
+                            [0.0, 1.0],
+                            [2.0, 3.0],
+                            [4.0, 5.0],
+                            [6.0, 7.0],
+                            [8.0, 9.0],
+                            [10.0, 11.0],
+                            [12.0, 13.0],
+                        ]
+                    ),
+                ],
+            )
```

Comment:
@cmarmo - it seems there is a problem here. This test is failing on the backport on v5.0.x (#14183, https://github.com/astropy/astropy/actions/runs/3715605304/jobs/6300973451), with an old version of Numpy. I investigated a bit, and found several problems:
- the test here should fail, it seems the data should be transposed. If I read `"test2.fits"` I get this:
```
In [3]: hdus = fits.open('test2.fits')

In [4]: hdus[1].data['test']
Out[4]: 
_VLF([array([[0., 1., 2., 3., 4.],
             [5., 6., 7., 8., 9.]]),
      array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.],
             [ 7.,  8.,  9., 10., 11., 12., 13.]])], dtype=object)
```
which is not the expected result. 
- I spend some time trying to find why the test was not failing on `main` until I realized, you are not testing the result of `np.array_equal` :grimacing: . The test code should use `assert np.array_equal` (or `np.testing.assert_array_equal`).
- the test case above (`test_empty_vla_raw_data`) is also missing `assert` in front of `np.array_equal`, and is also failing if I add the `assert`. This one is from  #13621.

I should have seen that so no blame on you, but it seems we now have a few issues to deal with.

Reply:
Also, the way the test is written raises a warning from Numpy: 
```
In [5]:             np.array_equal(
   ...:                 hdus[1].data["test"],
   ...:                 [
   ...:                     np.array(
   ...:                         [[0.0, 1.0], [2.0, 3.0], [4.0, 5.0], [6.0, 7.0], [8.0, 9.0]]
   ...:                     ),
   ...:                     np.array(
   ...:                         [
   ...:                             [0.0, 1.0],
   ...:                             [2.0, 3.0],
   ...:                             [4.0, 5.0],
   ...:                             [6.0, 7.0],
   ...:                             [8.0, 9.0],
   ...:                             [10.0, 11.0],
   ...:                             [12.0, 13.0],
   ...:                         ]
   ...:                     ),
   ...:                 ],
   ...:             )
/home/simon/.pyenv/versions/3.9.8/lib/python3.9/site-packages/numpy/core/numeric.py:2457: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  a1, a2 = asarray(a1), asarray(a2)
Out[5]: False
```
So I guess it would be better to test separately `hdus[1].data['test'][0)` and `hdus[1].data['test'][1]`, or make a `np.array(..., dtype=object)`.